File: day07/solution.exs
defmodule Parse do

  @doc """
  Group lines into instructions.
  """
  def chunkInstructions(line, command)

  # First line parsed
  def chunkInstructions(["$" | command], {"", []}), do: {:cont, {command , []}}

  # We have content and a new instruction.
  def chunkInstructions(["$" | command], acc), do: {:cont, acc, {command, []}}

  # For other stuff, it is the part of the current command.
  def chunkInstructions(content, {command, context}), do: {:cont, {command, [ content | context]}}

  def play(command, accumulator)

  # Go back to root doesn't change anything
  def play({["cd", "/"], _stdout}, {_cwd, structure, sizes}), do: {["/"], structure, sizes}

  # Go up a level ... but from root? Not allowed ...
  def play({["cd", ".."], _stdout}, acc = {["/"], _structure, _sizes}), do: acc

  # Go back up a level
  def play({["cd", ".."], _stdout}, {[_dir | parent], structure, sizes}), do: {parent, structure, sizes}

  # Go down a level
  def play({["cd", dir], _stdout}, {cwd, structure, sizes}), do: {[dir | cwd], structure, sizes}

  # Process listing
  def play({["ls"], stdout}, {cwd, structure, sizes}) do
    {size, subdirs} = stdout
      |> Enum.reduce({0, []}, fn elem, acc -> parse_ls(elem, acc) end)

    {cwd, Map.put(structure, cwd, subdirs),  Map.put(sizes, cwd, size)}
  end

  # Parse the output from ls.
  defp parse_ls(elem, acc)
  defp parse_ls(["dir", name], {size, subdirs}), do: {size, [name | subdirs]}
  defp parse_ls([file_size, _file], {dir_size, subdirs}) do
    {dir_size + String.to_integer(file_size), subdirs}
  end
end

defmodule Directories do
  @doc """
  Calculate directory total sizes.
  """
  def calc_sizes(cwd, structure, dir_sizes) do
    {c_total, c_map} = structure[cwd]
    |> Stream.map(fn subdir -> calc_sizes([subdir | cwd], structure, dir_sizes) end)
    |> Enum.reduce({0, %{}}, fn {e_total, e_map}, {a_total, a_map} -> {e_total + a_total, Map.merge(e_map, a_map)} end)

    dir_size = dir_sizes[cwd] + c_total
    {dir_size, Map.put(c_map, cwd, dir_size)}
  end
end

{_cwd, structure, sizes} = File.stream!("input", [:utf8], :line)
|> Stream.map(&String.split/1)
|> Stream.chunk_while({"", []}, &Parse.chunkInstructions/2, fn x -> {:cont, x, []} end)
|> Enum.reduce({[], %{}, %{}}, &Parse.play/2)
|> IO.inspect(label: "{cwd, structure, initial sizes}")

{_last_cwd, dir_sizes} = Directories.calc_sizes(["/"], structure, sizes)
|> IO.inspect(label: "{last_cwd, dir_sizes}")

# Part 1: Total size of directories under 100000
dir_sizes
|> Map.values()
|> Stream.filter(&(&1 <= 100000))
|> Enum.reduce(&(&1 + &2))
|> IO.inspect(label: "Part 1")

# Part 2: Find how much space we need to clear
# Step 1 - Work out how much space we need to clear, noting "/" is the total size
req_min = dir_sizes[["/"]] - (70000000 - 30000000)
# Step 2 - Find the smallest directory that is bigger than this requirement
dir_sizes
|> Map.values()
|> Enum.filter(&(&1 >= req_min))
|> Enum.min()
|> IO.inspect(label: "Part 2")

File: day07/solution.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Directory struct {
	parent  *Directory
	size    uint
	subdirs map[string]*Directory
}

type Accumulator struct {
	cwd  *Directory
	root *Directory
}

func check(err error) {
	if err != nil {
		log.Fatalf("Error: %v\n", err)
	}
}

func main() {
	file, err := os.Open("input")
	check(err)
	defer file.Close()

	root := Directory{subdirs: make(map[string]*Directory)}
	cwd := &root
	scanner := bufio.NewScanner(file)

	// Find the the first instruction
	scanner.Scan()

	// Parse the instructions
	for scanner.Scan() {
		line := scanner.Text()

		parts := strings.Split(line, " ")
		if parts[0] == "$" && parts[1] == "cd" {
			switch parts[2] {
			case "..":
				if cwd.parent != nil {
					cwd = cwd.parent
				}
			case "/":
				cwd = &root
			default:
				if cwd.subdirs[parts[2]] == nil {
					cwd.subdirs[parts[2]] = &Directory{
						parent:  cwd,
						subdirs: make(map[string]*Directory),
					}
				}
				cwd = cwd.subdirs[parts[2]]
			}
		} else {
			value, parse_err := strconv.Atoi(parts[0])
			if parse_err == nil {
				cwd.size += uint(value)
			}
		}
	}

	var sizes []uint = sizeStructure(&root)
	var totalSize = sizes[len(sizes)-1]

	required := totalSize - (70000000 - 30000000)

	var part1Total uint = 0
	var part2Size uint = 0

	for _, dirSize := range sizes {
		if dirSize < 100000 {
			part1Total = part1Total + dirSize
		}
		if dirSize >= required && (dirSize < part2Size || part2Size == 0) {
			part2Size = dirSize
		}
	}

	fmt.Printf("Part 1 total: %v\n", part1Total)
	fmt.Printf("Part 2 size:  %v\n", part2Size)
}

func sizeStructure(root *Directory) []uint {
	var total uint = root.size
	var sizes []uint = make([]uint, len(root.subdirs)+1)
	for _, info := range root.subdirs {
		subsizes := sizeStructure(info)
		sizes = append(sizes, subsizes...)
		total = total + subsizes[len(subsizes)-1]
	}
	return append(sizes, total)
}
